### ISSUE TITLE
Jax support for sparse matrices

### ISSUE THREAD

**Issue by GJBoth** [google/jax#6544]
Hi all,

I was discussing with @jakevdp in #6466 about adding support for sparse matrices in Jax. He already did quite some groundwork, but there's a lot of open questions on the what and how. Quoting Jake here:

> * 2D CSC/CSR/COO may not be the abstraction we want in the end... pydata-sparse has N-dimensional sparse arrays implemented using COO or a multi-dimensional generalization of CSR. Similarly taco has routines for general N-dimensional sparse computation. That route may be better, particularly for some deep learning workflows that e.g. have 4D batched arrays of sparse image sequences.
>* XLA scatter/gather implementations only go so far... any operations we implement should have efficient lowerings on as many backends as possible - I've started with cusparse on GPU because the low-level routines are already available in our GPU build process. This means that every non-trivial operation should be either implemented as a primitive or composed of other primitives
> * JAX is about composability, and in particular pmap/vmap, jit, and grad are central to the API. I want to spend some time exploring how we can best implement gradients of sparse operations or batched sparse operations, and also think about how ongoing work on dynamic shapes can be utilized to make jit of sparse operations more efficient and usable. With all that up in the air I want to keep the design minimal for now in case we need to re-think our approach based on what we find
> * all of the above should be motivated by real-world workflows that people have. I'm starting to gather a few example workflows so we can have that in mind.

So I set out to compare the implementations in pytorch and tensorflow to get a feel for what's out there, and considering the points above. Specifically, we're interested in how they implement it low-level: do they secretly turn everything dense under the surface, or are they given the full treatment with specific ops? 

I'm summarizing what I found in the table below  - TL;DR Pytorch offers better support. Since documentation and support are all over the place, in this [notebook](https://colab.research.google.com/drive/1dyfhD4SaSAfKPhx-aG7lF8BY88WlsWWx?usp=sharing) you can find a bunch of common operations with sparse tensor to see if they are supported. It's been a while since I did tensorflow, so let me know if I made a mistake!

| Feature                | Pytorch                                              | Tensorflow                                                   |
| :--------------------- | ---------------------------------------------------- | ------------------------------------------------------------ |
| Documentation          | [here](https://pytorch.org/docs/stable/sparse.html#) | [here](https://www.tensorflow.org/guide/sparse_tensor)       |
| Formats                | COO                                                  | COO                                                          |
| Supported ops          | variations of matmul                                 | sums, min, matmul, element-wise ops through `tf.sparse.map_values` |
| Specialized ops        | Probably[^5]                                                  | Seems so [^4]                                                         |
| Grad of sparse[^1]     | Limited to specific matmul                           | No.                                                          |
| Sparse grad[^2]        | Yes                                                  | No.                                                          |
| Dimensions             | 2D, Hybrid                                           | 2D - tensor supports higher-D, but operations not            |
| Uncoalesced[^3]allowed | Yes                                                  | No mention                                                   |
| Sparse - sparse matmul | No                                                   | No                                                           |
| Extra goodies | Has an adam optimizer for sparse tensor | Nope |

[^1]: Meaning gradient w.r.t to the sparse tensor
[^2]: Meaning calculating the gradient without turning it into a dense tensor
[^3]: Uncoalesced means duplicate entries in the coordinates - total value of the element is then sum of duplicate elements.
[^4]: Sparse ops are imported from gen_sparse_ops.py, which doesnt exist in the repo?
[^5]: They mention Torchcudasparse is now defunct [here](https://github.com/pytorch/pytorch/tree/master/aten/src) but I can't find any other mention in the source code.

To do
* Dig in the source code a little further (although if someone who actually is familiar with the code could chime in, that'd be great.)
* Check how pydata-sparse implement n-D sparse arrays.

Proposal:
* Get a very minimal differentiable sparse matmul working for COO format using XLA ops and one GPU-specific backend in cusparse (some help and guidance here would be nice)

I figured we can use this issue to track progress and discuss, so ... thoughts?

**Comment by jakevdp** [comment]
Looks great, thanks for tackling this! One comment: #6466 defines simple COO, CSR, CSC classes with implementations of sparse-dense matmul. We could start from there and think about autodiff rules.

A couple other notes:

- pydata-sparse has two ND sparse implementations: the main one is an N-dimensional COO, and the second that is under development is a generalization of CSR that essentially works by reshaping and flattening matrices into a standard 2D CSR representation.
- Another N-D sparse option is the one used by MLIR, with is a more general mix of sparse and dense dimensions that is a sort of generalization of CSR. I have a working JAX implementation of this that I'm currently cleaning up

**Comment by GJBoth** [comment]
Thanks! I'm working on implementing a jvp for the `coo_matvec` now, I'll make a PR soon. Quick API question: why did you choose the matvec to take in `rows, columns, data` instead of just giving the container and unpacking it? Isn't that much easier?

**Comment by jakevdp** [comment]
> why did you choose the matvec to take in rows, columns, data instead of just giving the container and unpacking it? Isn't that much easier?

When I wrote the primitive, I hadn't implemented any containers yet :grin:

**Comment by jakevdp** [comment]
In fact, I don't think the containers have been merged into master yet

**Comment by GJBoth** [comment]
Alrightie - I'll get on that too! As for not merging into master, thats okay - I forked Jax, pulled #6466 and made a new branch based of that. That seemed like the best option, is there a better one?

**Comment by jakevdp** [comment]
That's probably best for now... hopefully we'll merge #6466 soon.

Can you implement the AD rule without the container?

**Comment by GJBoth** [comment]
Yeah, I think so. We'll probably have to change it a little bit once we add them as there's no gradient w.r.t to cols and the rows, but we'll see about that then. Right now I'm just trying to get something working ;-)

**Comment by GJBoth** [comment]
I have to run some more tests but this little blurb seems to do the trick for the `coo_matvec`:

```
def _coo_matvec_jvp(primals_in, tangents_in, **params):
  vals, rows, cols, vec = primals_in
  mat_dot, _, _, vec_dot = tangents_in

  # Primal just calls coo_matvec
  primals_out = coo_matvec(vals, rows, cols, vec, **params)

  def make_zero(primal, tangent):
    return lax.zeros_like_array(primal) if type(tangent) is ad.Zero else tangent

  # Tangent is more complicated 
  # Jax requires the tangent vector v to have the same shape as the primal, which in
  # the sparse case is a vector with the non-zero values. Instead of turning mat_dot into a dense tensor and doing a dense 
  # matmul, we call the sparse matmul but with mat_dot instead of mat. 

  tangents_out = coo_matvec(make_zero(vals, mat_dot), rows, cols, vec, **params)
  tangents_out += coo_matvec(vals, rows, cols, make_zero(vec, vec_dot), **params)
  return primals_out, tangents_out
ad.primitive_jvps[coo_matvec_p] = _coo_matvec_jvp
```
I couldn't find much about about `ad.Zero` type except a short comment in the docs about adding your own primitives; is this the preferred way of handling it? Or is the proper way to check if it is zero and then not do the calculation at all? 

I'll run some more tests on this to confirm its correct, add an actual test and see if I can get the reverse mode working as well through the transposition,

**Comment by jakevdp** [comment]
#6466 is merged now, so the higher-level containers are in master.

Despite being compatible with Pytrees, these containers are not valid jaxpr types, and so we can't pass them directly to primitives like `csr_matvec_p`. We could modify the `csr_matvec()` and related functions to accept the higher-level objects, but I'd prefer to keep them closer to the primitive and use the class methods for higher-level operations.

The COO JVP rule looks really nice! I think the next step would be to add this to the primitive using the tools in `jax.interpreters.ad` (there are examples in `jax/_src/lax/lax.py`)

**Comment by GJBoth** [comment]
Great! I'll merge the changes with my branch.

> Despite being compatible with Pytrees, these containers are not valid jaxpr types, and so we can't pass them directly to primitives like csr_matvec_p. We could modify the csr_matvec() and related functions to accept the higher-level objects, but I'd prefer to keep them closer to the primitive and use the class methods for higher-level operations.

Would the solution I posted in the original issue be a possibility? I.e. we define a user facing API a la `csr_matvec(sparse_mat, vec)` which is simply a thin wrapper around `_csr_matvec(data, rows, vals, vec)` and possibly define a new `tree_flatten`. This could also help with the changing API; the user-facing would remain the same. 

> The COO JVP rule looks really nice! I think the next step would be to add this to the primitive using the tools in jax.interpreters.ad (there are examples in jax/_src/lax/lax.py)

For the COO JVP; I'll work on checking for ad.Zero. I guess the reason is so you don't need to propagate massive arrays of zeros through; is that correct? And thanks for the suggestion about adding it using `jax.interpreters.ad`, I'll look into it, but doesn't `ad.primitive_jvps[coo_matvec_p] = _coo_matvec_jvp` do that? 

I also think a good rule for working with sparse gradients would be that a sparse array has a sparse tangent. For example, `to_dense` would turn a sparse gradient into a dense one, and vice versa. I couldn't think of any examples where this rule wouldn't apply, WDYT?

**Comment by jakevdp** [comment]
I was thinking the user-facing API would be `sparse_mat @ vec`

**Comment by GJBoth** [comment]
That actually makes much more sense, but its complimentary to what I suggested, no? 

Anyway, I added jvp rules for the `coo_matmat`, `coo_todense` and `coo_fromdense`, meaning that all `coo functions` now have a jvp rule (and they can basically be copy-pasted to the CSR / CSC format). It uses sparse gradients wherever it can, and we can do some pretty funky stuff:
```
# Assume we have a dense matrix M_dense with nnz non-zero elements
M_sparse = coo_fromdense(M_dense, nnz=nnz)
y, dy = jax.jvp(lambda x: coo_matmat(*coo_fromdense(M_dense, nnz=nnz), coo_todense(data=x, row=M_sparse[1], col=M_sparse[2], shape=shape), shape=shape).sum(), (M_sparse[0], ), (jnp.ones_like(M_sparse[0]), ))
```
I'll add tests to `sparse_ops_test.py` this weekend and make a PR. Once that's done I was going to work on making it work with the new containers, or do you think something else would be more useful first (transposition rules for reverse-mode, more cusparse functions, ...)?

**Comment by jakevdp** [comment]
> That actually makes much more sense, but its complimentary to what I suggested, no?

Sure, we could make `coo_matvec()` and related functions accept the higher-level containers, but that's not possible for things like `_coo_matvec_impl` and `_coo_matvec_abstract_eval` because pytrees are not flattened into JAX objects at that level. I have a slight preference to have lower-level functions like `coo_matvec()` reflect the signature of their underlying implementations, but if you think it would be beneficial to change that I'm happy to go that route.

> I added jvp rules for the coo_matmat, coo_todense and coo_fromdense, meaning that all coo functions now have a jvp rule (and they can basically be copy-pasted to the CSR / CSC format).

Very cool! Looking forward to seeeing your PR!

**Comment by sunilkpai** [comment]
Hi @jakevdp what's the roadmap for sparse-sparse multiplication? I've seen one such implementation using autograd primitives in [ceviche](https://github.com/fancompute/ceviche/blob/master/ceviche/primitives.py) designed by a friend of mine, but was wondering if this might be available here in jax's experimental module at some point?

**Comment by jakevdp** [comment]
Work on sparse-sparse matmul is in progress in #7180

**Comment by jakevdp** [comment]
Experimental sparse support now has some documentation: https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html

**Comment by tetterl** [comment]
Awesome work here, already looking forward to give it a try!
Since we get more and more sparse support I was wondering if there are plans to support sparse direct solver (e.g. cholesky and LU solvers by using e.g. Pardiso and cuSOLVER)? It might be a bit tricky though to support an analytic and numeric factorization stage (maybe the sparsity pattern would need to be provided beforehand?).

**Comment by jakevdp** [comment]
I don't have any concrete plans for that, but it is definitely a possibility. In particular, cuSolver could be a compelling project because jaxlib already depends on CUDA, so it wouldn't require any new dependencies.

One related TODO is to lower BCOO routines to relevant cuSparse calls on the GPU; I haven't done this yet because I'm not entirely sure how to handle batched calls.

**Comment by zabboud** [comment]
> Has an adam optimizer for sparse tensor

I got here while looking for frameworks that support sparsity. I just wanted to add a small detail here - that actually pytorch's SparseAdam in its current implementation, works for a very specific use case. It does not work for Sparse Tensors, it actually requires the parameter tensor to be dense and the gradients associated with it to be sparse. There's an [issue](https://github.com/pytorch/pytorch/issues/101370) due to lack of clarity in the documentation on the use case. 

In terms of JAX - what is the current requirements/support for sparse gradient optimization?
